### ISSUE TITLE
Ota update when using rtos? 

### ISSUE THREAD

**Issue by Rhvarrier** [SensorsIot/ESP32-OTA#8]
Is this more a question than an issue. Why is ArduinoOta.handle() called only when Rtos is disabled? Is OTA not compatible with FREERTOS?

**Comment by 2010019970909** [comment]
No, it's simply because with FreeRTOS (and the define uncommented) the setup should launch the ArduinoOta.handle() in a task rather than in the loop. I did not test the code though.

**Comment by ReubenBTalbott** [comment]
I am unable to get RTOS to work as well, I did uncomment the RTOS option in the sketch. I am using platform IO with the Arduino framework so it isn't the Arduino IDE but it shouldn't affect it.

**Comment by 2010019970909** [comment]
After looking at the code closer (I did remark it before, I did not try, but I was suspecting some mistakes in the code). And indeed.

### Provide your credentials
First (it's not a mistake) you have to provide the ``credentials.h`` file (in the sketch folder directly or in the Arduino lib).
Personnally I did put it in the lib folder (``[...]\Arduino\libraries\Credentials``, ``[...]\Arduino`` is where your sketch are stored).

The text file ``credentials.h`` looks like:
```c++
#pragma once
const char* mySSID = "SSID of your AP";
const char* myPASSWORD = "Password of your AP";
```

### Analysis of the sketch and header and correction
I did first notice that the ``#define ESP32_RTOS`` was after the ``#include "OTA.h"``. It has to be inverted.

Secondly due to the previous point ("Provide your credentials"), the credentials are direclty used by the OTA.h header.
I don't really like that (as you may want to use directly some credentials in the main sketch). So, I'm including the ``credentials.h`` in the sketch and removing the includes in the header (as well as the ``const char* ssid, const char* password``, I prefer to pass them as arguments to the function ``setupOTA``).

Third, since we have moved the  ``#define ESP32_RTOS``  on top it should work... But there is two other little mistakes. First, the ``taskOne`` should be renamed to ``ota_handle``. And second, to have a working FreeRTOS task that function should loop and never return. You can achieve that with the following lines:

```c++
#if defined(ESP32_RTOS) && defined(ESP32)
void ota_handle( void * parameter ) {
  for (;;) {
    ArduinoOTA.handle();
    delay(3500);
  }
}
#endif
```

Fourth, I've changed the method in order to allocate dynamically the ``fullhostname`` in the setup function instead of imposing a max length.
### Full code modification
#### OTA_Template_Sketch.ino
```c++
#define ESP32_RTOS  // Uncomment this line if you want to use the code with freertos only on the ESP32
                    // Has to be done before including "OTA.h"

#include "OTA.h"
#include <credentials.h>

void setup() {
  Serial.begin(115200);
  Serial.println("Booting");

  setupOTA("TemplateSketch", mySSID, myPASSWORD);

  // Your setup code
  /*  Example
  pinMode(2, OUTPUT);
  */
}

void loop() {
#ifdef defined(ESP32_RTOS) && defined(ESP32)
#else
  ArduinoOTA.handle();
#endif

  // Your code here
  /* Example
  digitalWrite(2, HIGH);
  delay(500);
  digitalWrite(2, LOW);
  delay(1000);
  */
}
```

#### OTA.h
```c++
#ifdef ESP32
#include <WiFi.h>
#include <ESPmDNS.h>
#else
#include <ESP8266WiFi.h>
#include <ESP8266mDNS.h>
#endif

#include <WiFiUdp.h>
#include <ArduinoOTA.h>

#if defined(ESP32_RTOS) && defined(ESP32)
void ota_handle( void * parameter ) {
  for (;;) {
    ArduinoOTA.handle();
    delay(3500);
  }
}
#endif

void setupOTA(const char* nameprefix, const char* ssid, const char* password) {
  // Configure the hostname
  uint16_t maxlen = strlen(nameprefix) + 7;
  char *fullhostname = new char[maxlen];
  uint8_t mac[6];
  WiFi.macAddress(mac);
  snprintf(fullhostname, maxlen, "%s-%02x%02x%02x", nameprefix, mac[3], mac[4], mac[5]);
  ArduinoOTA.setHostname(fullhostname);
  delete[] fullhostname;

  // Configure and start the WiFi station
  WiFi.mode(WIFI_STA);
  WiFi.begin(ssid, password);

  // Wait for connection
  while (WiFi.waitForConnectResult() != WL_CONNECTED) {
    Serial.println("Connection Failed! Rebooting...");
    delay(5000);
    ESP.restart();
  }

  // Port defaults to 3232
  // ArduinoOTA.setPort(3232);

  // No authentication by default
  // ArduinoOTA.setPassword("admin");

  // Password can be set with it's md5 value as well
  // MD5(admin) = 21232f297a57a5a743894a0e4a801fc3
  // ArduinoOTA.setPasswordHash("21232f297a57a5a743894a0e4a801fc3");

  ArduinoOTA.onStart([]() {
    String type;
    if (ArduinoOTA.getCommand() == U_FLASH)
      type = "sketch";
    else // U_SPIFFS
      type = "filesystem";

    // NOTE: if updating SPIFFS this would be the place to unmount SPIFFS using SPIFFS.end()
    Serial.println("Start updating " + type);
  });
  
  ArduinoOTA.onEnd([]() {
    Serial.println("\nEnd");
  });
  
  ArduinoOTA.onProgress([](unsigned int progress, unsigned int total) {
    Serial.printf("Progress: %u%%\r", (progress / (total / 100)));
  });
  
  ArduinoOTA.onError([](ota_error_t error) {
    Serial.printf("Error[%u]: ", error);
    if (error == OTA_AUTH_ERROR) Serial.println("\nAuth Failed");
    else if (error == OTA_BEGIN_ERROR) Serial.println("\nBegin Failed");
    else if (error == OTA_CONNECT_ERROR) Serial.println("\nConnect Failed");
    else if (error == OTA_RECEIVE_ERROR) Serial.println("\nReceive Failed");
    else if (error == OTA_END_ERROR) Serial.println("\nEnd Failed");
  });

  ArduinoOTA.begin();

  Serial.println("OTA Initialized");
  Serial.print("IP address: ");
  Serial.println(WiFi.localIP());

#if defined(ESP32_RTOS) && defined(ESP32)
  xTaskCreate(
    ota_handle,          /* Task function. */
    "OTA_HANDLE",        /* String with name of task. */
    10000,            /* Stack size in bytes. */
    NULL,             /* Parameter passed as input of the task */
    1,                /* Priority of the task. */
    NULL);            /* Task handle. */
#endif
}
```
I've tested it and it seems to work on the ESP32 and ESP8266 (without FreeRTOS). I won't do a push request since there is a lot of changes and some personnal choses.

**Comment by 2010019970909** [comment]
> I am unable to get RTOS to work as well, I did uncomment the RTOS option in the sketch. I am using platform IO with the Arduino framework so it isn't the Arduino IDE but it shouldn't affect it.

If you are using the ESP8266, it won't work.

**Comment by ReubenBTalbott** [comment]
I am using an ESP32 so RTOS should work, I will try out your modifications, really hope I can get it to work 😅

Thanks!

**Comment by 2010019970909** [comment]
@ReubenBTalbott, I hope so. Keep me updated.

You are welcome ;-)

**Comment by ReubenBTalbott** [comment]
It worked! Thanks a lot! Since you don't want to push the modified code to the repo perhaps consider creating a fork?

**Comment by 2010019970909** [comment]
Hello all,

@ReubenBTalbott, noted. I did not test the Telnet sketch but I did update the code, and I've tried to take the Telnet feature into account. I've also added some information to the readme to make it a little bit easier to use. If you try it out, give me your feedback 😉 

@Rhvarrier, maybe you can close this issue since, we did reply to your questions.

@SensorsIot, if needed, I can do a push request, but as said above, I have made some personnal choices and I did not fully test the code.

Goodbye 😃

**Comment by SensorsIot** [comment]
@Vincent: Yes, please create a pull request

**Comment by Bach-Danielsson** [comment]
Thanks for your contribution.  I use platformIO and get this code working on my ESP32-S. But i cant get the version with TelnetStream working. 
Im abel to connect by Putty , but no text is displayed.


void loop()
{

#ifdef defined(ESP32_RTOS) && defined(ESP32)
#else
  ArduinoOTA.handle();
#endif

  // Your code here
  entry = micros();
  i = i + 1;
  Serial.print("LoopteleStream_Serial port ");
  Serial.println(i);
  delay(1000);
  TelnetStream.print(micros() - entry);
  TelnetStream.println("  LoopteleStream_TelnetStream");

  delay(1000);
  /* Example
  digitalWrite(2, HIGH);
  delay(500);
  digitalWrite(2, LOW);
  delay(1000);
  */
}

**Comment by 2010019970909** [comment]
@Bach-Danielsson,

I did not try to use the Telnet version. Apparently it doesn't work with the ``#define ESP32_RTOS`` and I don't know why... (at least it's not obvious for me). I did manage to use the code as it is by commenting the ``#define ESP32_RTOS`` and using the version 1.0.0 of the [TelnetStream](https://github.com/jandrassy/TelnetStream) library. I was able to use their example alone, so there might be some interactions between  OTA and the TelnetStream and maybe even between the ``#define ESP32_RTOS`` functionnality and the TelnetStream. To give more visibility to the issue, you can open a new issue here or/and on [TelnetStream](https://github.com/jandrassy/TelnetStream).

Regards,
Vincent
